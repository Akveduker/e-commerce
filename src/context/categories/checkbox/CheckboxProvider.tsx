import { createContext, FC, useContext } from "react";
import { CheckboxNames } from "../../../types/filter/filter";
import React from 'react';


export const CheckboxContext = createContext<CheckboxNames | null>(null)
interface CheckboxProviderProps {
    checkboxName: CheckboxNames;
    children: React.ReactNode;
}
const CheckboxProvider: FC<CheckboxProviderProps> = ({ checkboxName, children }) => {
    return (
        <CheckboxContext.Provider value={checkboxName}>
            {children}
        </CheckboxContext.Provider>
    );
};

export default CheckboxProvider;