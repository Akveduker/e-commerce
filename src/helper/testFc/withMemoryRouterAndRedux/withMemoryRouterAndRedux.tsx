import { configureStore } from "@reduxjs/toolkit"
import { render } from "@testing-library/react"
import { InitialEntry } from "history"
import { Provider } from "react-redux"
import { MemoryRouter } from "react-router-dom"
import { mockReducersForStore } from "../../mock/mockStore/mockStore"
export const withMemoryRouterAndRedux = (children: React.ReactNode, reducers?: Partial<typeof mockReducersForStore>, initialEntries?: InitialEntry[]) => {
    const mockStore = configureStore({
        reducer: { ...mockReducersForStore, ...reducers }
    });
    return render(
        <MemoryRouter initialEntries={initialEntries}>
            <Provider store={mockStore}>

                {children}
            </Provider>
        </MemoryRouter>
    )
}