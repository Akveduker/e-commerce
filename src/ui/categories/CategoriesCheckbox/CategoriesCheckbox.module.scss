@import "../../../styles/_vars.scss";
.checkbox {
    display: none;
}
.fake-checkbox {
    border: 1.5px solid $C1D;
    width: 24px;
    height: 24px;
    display: block;
    position: relative;
    border-radius: 4px;
    &::after {
        content: "";
        background: url("../../../assets/icons/categories-checkbox-true.svg") no-repeat center;
        display: block;
        position: absolute;
        width: 100%;
        height: 100%;
        top: 0;
        left: 0;
        opacity: 0;
    }
}
.label {
    display: block;
    &--active {
        .fake-checkbox {
            border-color: $C2B;
            background-color: $C2A;
            &::after {
                opacity: 1;
            }
        }
    }
}
