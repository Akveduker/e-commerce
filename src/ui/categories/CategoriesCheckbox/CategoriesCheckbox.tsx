import React, { FC, useContext, useState } from 'react';
import { CheckboxContext } from '../../../context/categories/checkbox/CheckboxProvider';
import { useCheckboxDispatch } from '../../../hooks/categories/filters/checkbox/useCheckboxDispatch';
import { useAppSelector } from '../../../store/store';
import s from './CategoriesCheckbox.module.scss'
interface CategoriesCheckboxProps extends React.InputHTMLAttributes<HTMLInputElement> {
    checkId: number;
}
const CategoriesCheckbox: FC<Omit<CategoriesCheckboxProps, 'type' | 'className' | 'checked'>> = ({ checkId }, props) => {
    const name = useContext(CheckboxContext)
    const currentItem = useAppSelector(state => state.filters.checkboxes.find(checbox => checbox.name === name))
    let status = currentItem?.activeIdArr.includes(checkId) ? true : false
    const [value, setValue] = useState(status)
    const [checkedTrue, checkedFalse] = useCheckboxDispatch(checkId)
    const checkboxChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const checked = e.target.checked
        setValue(checked)
        checked ?
            checkedTrue()
            :
            checkedFalse()
    }
    return (
        <label className={value ? `${s['label']} ${s['label--active']}` : s['label']} data-testid={'label'}>
            <input type="checkbox" className={s['checkbox']} onChange={checkboxChange} checked={value} {...props} />
            <span className={s['fake-checkbox']}></span>
        </label>
    );
};

export default CategoriesCheckbox;