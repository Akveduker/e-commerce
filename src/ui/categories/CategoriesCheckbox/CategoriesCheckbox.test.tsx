import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import CheckboxProvider from "../../../context/categories/checkbox/CheckboxProvider"
import { withMemoryRouterAndRedux } from "../../../helper/testFc/withMemoryRouterAndRedux/withMemoryRouterAndRedux"
import CategoriesCheckbox from "./CategoriesCheckbox"

describe('CategoriesCheckbox', () => {
    /* чекбокс используется с именем brands потому что в инициализации фекового стора 
    указан обьект с этим именем
    */
    const element =
        <CheckboxProvider checkboxName='brands'>
            <CategoriesCheckbox checkId={1} />
        </CheckboxProvider>
    test('снапшот неактивного чекбокса', () => {
        const { container } = withMemoryRouterAndRedux(element)
        expect(container).toMatchSnapshot()
    })
    test('Смена состояния работает, класс поменялся', () => {
        withMemoryRouterAndRedux(element)
        const label = screen.getByTestId('label')
        userEvent.click(label)
        expect(label).toHaveClass('label--active')
    })
    test('снапшот активного чекбокса', () => {
        const { container } = withMemoryRouterAndRedux(element)
        const label = screen.getByTestId('label')
        userEvent.click(label)
        expect(container).toMatchSnapshot()
    })
})