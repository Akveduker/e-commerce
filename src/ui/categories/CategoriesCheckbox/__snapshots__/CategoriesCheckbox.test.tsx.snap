// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesCheckbox снапшот активного чекбокса 1`] = `
<div>
  <label
    class="label label--active"
    data-testid="label"
  >
    <input
      class="checkbox"
      type="checkbox"
    />
    <span
      class="fake-checkbox"
    />
  </label>
</div>
`;

exports[`CategoriesCheckbox снапшот неактивного чекбокса 1`] = `
<div>
  <label
    class="label"
    data-testid="label"
  >
    <input
      class="checkbox"
      type="checkbox"
    />
    <span
      class="fake-checkbox"
    />
  </label>
</div>
`;
